### DISCUSSION TITLE
How to highlight text in the node core

### DISCUSSION THREAD

**Discussion by euu2021** [freeplane/freeplane#1271] [General]
I see that the bugs/feature requests page in the old forum is not available anymore. If possible, please make it visible again. 

I remember that, once, I wrote, on a ticket page, some tips about how to highlight text in Freeplane. Now, I need the tips myself, but I can't remember what I wrote there 😅

By the way, do you have any ideas about how to solve this problem: I need to highlight (I mean, use the highlighter tool to format the background of the word as yellow) a piece of text in the node core. I want to highlight directly in the node core, ie, when editing the text in-line, without the need to open a second window.

Thanks,

**Comment by euu2021** [comment]
I found a workaround: changing the CSS of Bold, Underline and Italics. Those 3 options can already be applied directly to the inline text, so the workaround is to "replace" them with a highlighter appearance. For example, including this into the CSS:

```
u {
background-color: yellow;
}

b {
background-color: blue;
}

i {
background-color: green;
}
```

The downside is that the user loses the option to use the actual Bold, Underline and Italics.

**Reply by euu2021** [reply]
This workaround doesn't work with minimized texts, as explained at #1272, which is essential to me. So, I'm back to square one.

The idea that I proposed in the past had something to do with Autohotkey, but I can't remember what it was 🤣🤣

**Comment by macmarrum** [comment]
@euu2021 How would you highlight a piece of text in a dialog window (SimpyHTML)?

**Reply by euu2021** [reply]
Using a dialog, I see two options:

-SimplyHTML has an option to paste html, so I would create a AHK script that grabs the selected text, transform it (insert the highlight tags), then paste it back as html

-or use this script by Edo, in order to open the edit the text in an editor that has a highlighting tool: #408 

The problem with using a dialog is that it's slower than formatting directly in the node. When highlighting a lot of text, it's not practical.

**Reply by euu2021** [reply]
to picture more clearly the difference between editing in-line, and editing in a dialog:
As I'm working on a lot of nodes full of text, I go through the nodes doing those steps:

-1: I go through the nodes, checking if there is something to be done on it
-2: If there is something to do on the node, then I open it as "edit-in line". Notice that, with the edit in line, the user can edit the text on the spot, without visually changing the screen; while, opening a dialog the screen changes completely, and the user loses focus of where he was looking at, which breaks the flow.
-3: work on the text: rewrite, erase, add text, split, bold, underline, (insert highlight here)
-4: after I finish, I press enter, and I'm ready to go on to the next node, because my eyes are already there, on the correct spot

**Comment by dpolivaev** [comment]
I made it visible.

**Reply by euu2021** [reply]
Thanks!

**Comment by euu2021** [comment]
[This is](https://sourceforge.net/p/freeplane/featurerequests/2575/) the page that I was looking for.

My idea there doesn't solve the problem, though. It relies on using Markdown on the node, but Markdown doesn't allow in-line edit. So, it is the same of opening on a new dialog.

**Reply by amariusz** [reply]
Maybe option `Preferences->Plugins->Markdown->EDit Markdown in-line` would be useful?

**Reply by euu2021** [reply]
Nice! thanks

___
So, using Markdown is an option. But, I think it is still not a good option, because the syntax is too lengthy, making it difficult to edit the text:
![image](https://github.com/freeplane/freeplane/assets/77707706/5c396aa8-1c89-4a5a-8d9b-69a6cdb221a2)

**Reply by amariusz** [reply]
IMO the options are very limited. You can achieve shorter syntax with `<b>`, `<i>` and `<u>` as you mentioned above. It seems they don't conflict with markdown bold/italic syntax (with stars or underscores).  You can try a similar approach with `<pre>` tag. If you need more formatting options then you can define CSS classes and use `<span class='abc'>` which is a little shorter than `style` but it's noisy already.

**Reply by euu2021** [reply]
Thanks for the suggestions. 

Yes, Markdown has this downside of making things difficult to read in the edit mode. My brain is not ready for that 🤣🤣

`<b>, <i> and <u>` used in Markdown show the same problem that I reported with html: #1272

**Reply by amariusz** [reply]
There are markdown flavors that support highlighting with `==highlighted text==` syntax. Similarly strikethrough with `~~crossed out text`. Obsidian is one example: https://help.obsidian.md/Editing+and+formatting/Basic+formatting+syntax#Styling+text

Having this support in FreePlane could be useful. I would +1 such feature request ;)

**Comment by dpolivaev** [comment]
Control+R, Control+G and Control+L make selected characters Red, Green and BLue in inline editors and in dialog editors.
Also in the dialog editor and in note pane you can change character color from Format->Font.

Unfortunately, the background can not be changed like that because the framework does not know how to properly save this formatting.

**Reply by euu2021** [reply]
I see. Thanks.

**Comment by dpolivaev** [comment]
I decided to try to rewrite parts of SimplyHTML so that `<span>` element and all related attributes including the background color for characters can be supported.

**Reply by euu2021** [reply]
Nice! It will be useful. Text highlighters really help in the formatting. Thanks

**Reply by macmarrum** [reply]
That's great!
I keep my fingers crossed :crossed_fingers:

**Reply by dpolivaev** [reply]
The feature is available in preview 1.11.5_17

**Reply by euu2021** [reply]
> The feature is available in preview 1.11.5_17

![image](https://github.com/freeplane/freeplane/assets/77707706/115f4899-8af5-4ed7-8642-2d125035972c)

**Comment by euu2021** [comment]
(I don't know if the following idea makes any sense, technically. If it doesn't, please ignore it 😅)

I see that the Highlighter is implemented as an inline style. Now, I'm wondering if it wouldn't be interesting to have it as an external style sheet (i.e., as a span class). 

This would allow the user to change the color of each highlighter by changing the characteristics of the class in the CSS. 

In my situation, for example, I use a different background color for each node style. So, for each node style I would have a different CSS, in order to have the highlighted text always looking good, no matter the node style.

It seems like a way of giving more flexibility, and I don't see apparent drawbacks.

**Reply by euu2021** [reply]
Now, I see that the drawback happens when exporting, because the style is not in the text. So, never mind.

**Comment by i-plasm** [comment]
Hello! I recently published __SimplyClassyCSS__ (https://i-plasm.github.io/simply-classy-css/), which supports really easy highlighting in Freeplane Markdown (core, details, and/or note). It includes the use of scripts for automatic highlighting. I'll allow the images below to speak for themselves! 

(___IMPORTANT! Freeplane versions v1.11.3-v.1.11.6 are incompatible with SimplyClassyCSS. The incompatibilities seem to have been solved in v1.11.7. At least Freeplane v1.9.12 is needed.___)

![testing_markdown_freeplane_simplyclassycss790X535.png](https://raw.githubusercontent.com/i-plasm/simply-classy-css/main/docs/testing_markdown_freeplane_simplyclassycss790X535.png)

![example3](https://github.com/freeplane/freeplane/assets/144830522/38c27f82-056e-4f57-acc1-8b5d9786c0bb)

To respond to a concern of yours: Although the highlighting is based on CSS classes, the corresponding tags will be embedded in the Markdown content, so they won't be lost when exporting (at least in the export methods I've tried). It's best that they be embedded otherwise you would have to set CSS styles to go which each predefined node core style (each defining the default `<p>` so that it acts as a highlighter), and managing more than a couple of CSS styles  in the same map can become messy, AND you wouldn't have the ability to use different highlighter colors inside a single node core/details/note. And of course, the highlighter info would be lost at export. All of this is solved with SimplyClassyCSS: __When exporting a map that uses SimplyClassyCSS highlighting tags, it's only a matter of a simple search and replace, and the highlighting html tags will be understood anywhere without the use of SimplyClassyCSS__. A simple search of __`span class="simply-`__ and a replace with __`span style="background-color:`__ suffices to transform the exported tags to 'default' html.

Highlighting is very convenient with scripts: 
1. the user selects the text to be highlighted, 
2. calls the script (faster via a shortcut), 
3. and the selected text will be wrapped with the special tag.

You may check:

- __SimplyClassyCSS__: https://i-plasm.github.io/simply-classy-css/
- __Detailed guide for highlighting in Freeplane__: https://ideaplasm.com/2023/09/29/freeplane-highlighting-emphasizing/

I hope this helps!

**Reply by macmarrum** [reply]
@i-plasm This is an interesting idea. I can see a lot of thought put into it. I can imagine people using this approach, especially if they work with Markdown content in their mind maps. I suppose that if someone only needs text highlighting, they will probably use the built-in (HTML) highlighting feature, which is much simpler to apply.
Did you consider to create mind map templates which already contain your CSS, maybe one template per an example from your [Gallery](https://i-plasm.github.io/simply-classy-css/site/gallery.html)? Alternatively, a script or even an add-on that installs your CSS into the current mind map, maybe from a selection of themes, like in your Gallery?

**Reply by euu2021** [reply]
Interesting! I like the flexibility of using CSS classes.

Now, there is no way to make the SpanHighlighter script work with HTML nodes, right? I mean, when editing the node core in-line, it pastes the text as plain text, so it ends up like that:

![image](https://github.com/freeplane/freeplane/assets/77707706/8289e11e-c6db-4765-a316-cefcc72aa785)

**Reply by i-plasm** [reply]
@macmarrum , thank you for your comment. I agree, if the user doesn't need markdown, they are well served with the new highlight feature in the HTML editor. No need to handle html with scripts! 

In my case, working with markdown is so much faster that I've tried to circumvent its limitations as much as I can. With highlighting, once the user installs the _SpanHighlighter_ script, it's just a matter of selecting the text in the markdown editor and calling the script (faster via a shortcut!). That works really well in my experience, and no harder than if you were working with a visual html editor. 

I think the most cumbersome part of this approach is the first stage: assigning a SimplyClassyCSS style to whatever you want to become markdown-highlightable, i.e core, note, details, and/or predefined styles. Hopefully the user configures the map as per their needs, and saves it as a template, so this CSS-assigning process wouldn't have to be repeated for every new mindmap. With such a template, this stage is never required again. Your suggestion to make template mindmaps that come with "preinstalled" SimplyClassyCSS is great! I definitely am considering that.

I've actually considered - and in fact built a few months ago -a plug-in for assigning CSS styles easily and intuitively. It comes preloaded with CSS styles, and all the user has to do is choose the CSS from a dropdown menu. However, I haven't completed it because just to assign a CSS to let's say the notes or details, I had to dig into really internal Freeplane code (and I'm talking about code from v1.9.12). And I could not find any way to set CSS on predefined node styles - that would be really important -. I would need Freeplane public API to allow me two things: 1) a way to access the list of predefined node styles (at least I didn't find any way to do so) and 2) to allow setting the CSS reliably to any such predefined styles. With my plug-in experiment, I am able to successfully set the CSS on certain elements (note, details, individual cores) but then for some reason Freeplane doesn't register the setting of CSS as a map change, so when I close Freeplane it doesn't ask "Do you want to save the mindmap...?" and the CSS changes get lost. So new methods in the public API addressing all those issues would be fundamental.  This is how the plug-in currently looks like:

![css-early-plugin-freeplane-atom-dark](https://github.com/i-plasm/HelloWorld/assets/144830522/38f757e9-8418-4f9d-8bec-88b8d1625bb0)

@euu2021  sorry about the lack of clarity of my earlier response. Glad you figured it out!

>  there is no way to make the SpanHighlighter script work with HTML nodes, right?

EDIT: to recommend that the highlighting method I proposed only be used for Markdown. I tested different Freeplane versions, and all of them except v.1.11.7 seem to rewrite the HTML and strip it from the special CSS classes. Therefore the highlighting is lost in HTML content type in Freeplane v1.11.6 and lower. I've never seen that problem when using Markdown content type,

<del> For the current version of SpanHighlighter to work with HTML nodes, unfortunately it cannot be done in-line, the user would have to switch to the "HTML code view" tab of the HTML Editor, and selecting/manipulating the text to be highlighted would be too cumbersome and error-prone if one isn't careful. The access to SimplyHTML's editor panes is almost totally locked for groovy scripts. So, as of right now, the scripts are best suited for markdown nodes. If I come up with a solution I'll update the scripts. </del>

This is a picture of how the markdown editor looks like with the special highlighting tag and the output (note: the scripts also work in the in-line markdown editor) :

![markdowneditor-simplyclassycss](https://github.com/i-plasm/HelloWorld/assets/144830522/ec5a35fb-b575-408c-b362-f3ffe623107f)

**Reply by euu2021** [reply]
The `Edit->Node core->Edit node core in dialog` has a "paste as HTML" option, so one (cumbersome) solution that I can think is an autohotkey script that does something like that:

- the user selects a piece of text to be highlighted in the in-line editor
- the script automatically opens the `Edit node core in dialog`, 
- highlight the text selected by the user
- and closes the editor

**Reply by i-plasm** [reply]
@euu2021 right, there can be workarounds like that, but cumbersome as you say. But I have two good news: first, today I solved the issue for in-line highlighting of HTML-node core (at least in my tested environment) . Worked really great in my experience and I hope to share the script and instructions perhaps this weekend. Easy and fast! (select the text in-line, invoke the script via a shortcut and the highlight appears immediately).

Second good news: today I discovered that Freeplane v1.11.7 (latest version currently) allows in-line HTML core highlighting as well! I discovered it because I typed ` ctrl shift L ` by accident, and that's the shortcut for Freeplane's blue highlighter! So, totally by accident. The available shortcuts are found in the SimplyHTML external editor, in its "__Format__" menu. A big difference between my highlighter script and Freeplane's is that the colors are hard-coded in Freeplane and only look good with very dark fonts, whereas with the SimplyClassyCSS method, the highlights always look good no matter what SimplyClassyCSS you choose. But obviously Freeplane's highlighter doesn't need any extra  set-up, so both would have pros and cons. This is what I'm referring to regarding the colors issue:
![freeplane-simplyclassycss-highlighter-colors](https://github.com/freeplane/freeplane/assets/144830522/3bcb77d5-ec36-4318-9056-8652adb6fbe1)

Best regards

**Reply by euu2021** [reply]
> But I have two good news: first, today I solved the issue for in-line highlighting of HTML-node core (at least in my tested environment) . Worked really great in my experience and I hope to share the script and instructions perhaps this weekend. Easy and fast! (select the text in-line, invoke the script via a shortcut and the highlight appears immediately).

I'm curious about how this script works.

> A big difference between my highlighter script and Freeplane's is that the colors are hard-coded in Freeplane and only look good with very dark fonts, whereas with the SimplyClassyCSS method, the highlights always look good no matter what SimplyClassyCSS you choose.

The FP highlighter automatically adapts (dark or light) when the text is highlighted:
![image](https://github.com/freeplane/freeplane/assets/77707706/a5d4e31b-566d-4baf-8be7-8a38424b7984)


But, it doesn't change afterward:
![image](https://github.com/freeplane/freeplane/assets/77707706/e72ae3c8-1074-4bbe-87cc-953f4e626df3)


 So, using classes would be more flexible, indeed.

**Reply by pmking369** [reply]
Hi Is there a way to find and highlight text within a node.. like find and replace?
Any suggestion how to achieve that

**Reply by euu2021** [reply]
> Hi Is there a way to find and highlight text within a node.. like find and replace?
> Any suggestion how to achieve that

If I understood correctly, your question is related to that request: https://github.com/freeplane/freeplane/issues/1650
